### ISSUE TITLE
[HOLD for payment 2023-04-27] [$1000] Date field allows spaces between the numeric digits, unlike other numeric fields (like Rate)

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#17280]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Go to web Chrome/Android
2. Go to any workspace
3. Go to reimburse expenses and try to add a space in the rate field. Notice that it doesn't allow any spaces
4. Now go to the personal information and date of birth field
5. Go to year and try to input a number followed by spaces and notice that it allows spaces in the field

## Expected Result: 
Space should not be allowed in the year field - it should mirror that of the rate input field.

## Actual Result: 
Spaces are allowed in the year search field between the numeric digits. 

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.98-2 / v99-4
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
<img width="960" alt="Untitled" src="https://user-images.githubusercontent.com/43996225/231254181-e7308adc-7813-4ace-8f30-a1050ed5d2ff.png">

https://user-images.githubusercontent.com/43996225/231254245-fdc104a2-d8b3-4b2c-bb7d-22fc4756c3ab.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @priya-zha
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1681201745994509

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01ecf6260853d6f525</li>
        <li>Upwork Job ID: 1645992955077120000</li>
        <li>Last Price Increase: 2023-04-12</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @jliexpensify (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by jliexpensify** [comment]
On v99.4 (Mac) and can reproduce. Also occurs on Android (Pixel 3a, v98-2).

This is a minor bug, but I agree that the year field should mirror the rate field, as there aren't spaces in a year.

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01ecf6260853d6f525

**Comment by MelvinBot** [comment]
Triggered auto assignment to @trjExpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @s77rt (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @hayata-suenaga (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by dhairyasenjaliya** [comment]
### Proposal

**Please re-state the problem that we are trying to solve in this issue.**

- Date field allows spaces between the numeric digits, unlike other numeric fields (like Rate)

**What is the root cause of that problem?**

- The filterYearList function in the YearPickerPage class does not filter out spaces from the input text, resulting in spaces being allowed in the year search field.

**What changes do you think we should make in order to solve the problem?**

- We are trimming the values when we are filtering the year data only the thing is we are not validating the user-typed number so for that we just need to make sure white space is replaced with the blank which is a simple regex to validate `replace(/ /g, '')` we are already using in `modifyRoomName()` , validating the debit card number , emoji utils 

- We dont need to add regex `\s/g` (which is used to remove whitespace from a string (including spaces, tabs and newline characters ) simply because the user never gets a chance to add `tabs and newline characters` **I.e** -> when the user press tab it changes focus and we are already limiting the value to 4 characters to the max and number of line is 1 so it doesn't make sense to add regex `\s/g`

- Since we are filtering displayed value that we can also add to `yearOptions` as well so need to remove `text.trim()`

```diff

    filterYearList(text) {
+        const filteredText = text.replace(/ /g, '');
        this.setState({
+            inputText: filteredText,
-            yearOptions: _.filter(this.yearList, year => year.text.includes(text.trim())),
+           yearOptions: _.filter(this.yearList, year => year.text.includes(filteredText)),
        });
    }

```

**Note**

- IMO we should not replace text with blank space `([\D]/g` its simply a bad UX behavior and also we never used it anywhere else in the app
 
- We are already limiting the text character to a maximum of 4 so there is no need to add validation here 

@s77rt

**Comment by gargmegham** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The current implementation allows spaces, and characters in the year text input of YearPickerPage. But it should only allow numeric characters

What is the root cause of that problem?
The keyboard type number-pad is not working as expected, because it should have only allowed numbers by default

What changes do you think we should make in order to solve the problem?
Before modifying the state inside the filterYearList function input text should be formatted just like we're doing in rate input field mentioned in issue description

Eg.
```
filterYearList(text) {
        // Remove all non-numeric characters from the input
        const inputYear = text.replace(/[^0-9]/g, '');
        this.setState({
            inputText: inputYear,
            yearOptions: _.filter(this.yearList, year => year.text.includes(inputYear)),
        });
    }
```

**Comment by gargmegham** [comment]
> 📣 @gargmegham! 📣
> 
> Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork. Please follow these steps:
> 
> 1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
> 2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
> 3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.
> 
> ![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)
> 
> Format:
> 
> ```
> Contributor details
> Your Expensify account email: <REPLACE EMAIL HERE>
> Upwork Profile Link: <REPLACE LINK HERE>
> ```
Contributor details
Your Expensify account email: meghamgarg@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~0170461d5a03bcee8d

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by s77rt** [comment]
@hayata-suenaga Please remove the `Reviewing` label. No one is assigned yet.

**Comment by s77rt** [comment]
@Prince-Mendiratta Thanks for the proposal. Your RCA is correct and the fix looks good to me :+1:.

:ribbon: :eyes: :ribbon: C+ reviewed

cc @hayata-suenaga

**Comment by s77rt** [comment]
@dhairyasenjaliya Thanks for the proposal. The RCA is correct but it's basically a copy-paste even if you agree with someone's RCA you should at least rephrase or give credit to the original writer. As for your concern about replacing other text, we do that already [here](https://github.com/Expensify/App/blob/8adee6bdf733058985e55ba3fc0d39f72cddc5cf/src/pages/workspace/reimburse/WorkspaceReimburseView.js#L147).

**Comment by dhairyasenjaliya** [comment]
Hm oky point noted was simple RCA thank you @s77rt

**Comment by s77rt** [comment]
@Prince-Mendiratta 
> What are your thoughts on additional constraints?

The length is already limited to 4 [here](https://github.com/Expensify/App/blob/8adee6bdf733058985e55ba3fc0d39f72cddc5cf/src/pages/YearPickerPage.js#L93).

**Comment by s77rt** [comment]
@hayata-suenaga before assigning @Prince-Mendiratta do you think this should be handled by @ArekChr as a part of https://github.com/Expensify/App/issues/14322? We had a discussion about this bug but we decided to ignore it at that time, I think mostly because we didn't really consider it as a bug and I was not really convinced about the way it was meant to be handled.

I think we should hire @Prince-Mendiratta here since the solution we were going with is slightly different.

Just stating the above for clarity and transparency. cc @ArekChr
.

**Comment by hayata-suenaga** [comment]
@s77rt thank you for the proposal review.

I think we can go with @Prince-Mendiratta. Assigning @Prince-Mendiratta

**Comment by MelvinBot** [comment]
📣 @Prince-Mendiratta You have been assigned to this job by @hayata-suenaga!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by MelvinBot** [comment]
@trjExpensify, @s77rt, @Prince-Mendiratta, @hayata-suenaga Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by s77rt** [comment]
Not overdue. PR deployed to staging 19 hours ago.

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.3.2-5 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/17356


If no regressions arise, payment will be issued on 2023-04-27. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by trjExpensify** [comment]
Hm, where's the checklist for this? Adding it manually.. 

**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@s77rt] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@s77rt] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@s77rt] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@trjExpensify] Determine if we should create a regression test for this bug.
- [ ] [@s77rt] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@trjExpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by trjExpensify** [comment]
Oh, maybe. I didn't think that had an impact tbh. Payments check in the meantime. I can see the PR was approved on the 13th, so within the timeframe, the slow down on merge was on our side. 

- $1,500 to @Prince-Mendiratta for the fix & #urgency bonus
- $1,500 to @s77rt for C+ & #urgency bonus 
- $250 to @priya-zha for the bug report. 

Sound about right?

**Comment by trjExpensify** [comment]
>I don't think bonus applies here, this was merged on 18th.

Yeah, I looked at that, but you were both done and approved on the 13th. After the weekend, the hold up was on us internally to action on the merge. Do you think I'm missing something?

**Comment by s77rt** [comment]
- The PR that introduced the bug has been identified: Not exactly a bug. This more like a feature request / improvement. This may have been handled on https://github.com/Expensify/App/issues/14322 but didn't as explained on https://github.com/Expensify/App/issues/17280#issuecomment-1504735586.
- The offending PR has been commented on: I don't think this applies here since it's not a regression. cc @hayata-suenaga
- A discussion in #expensify-bugs has been started: Same ^
- Determine if we should create a regression test for this bug: Yes

___

**Regression Test Proposal**

1. Open App
2. Go to Settings > Profile > Personal Details > Date of birth
3. Click on the year
4. Verify you are redirected to the Year picker page
5. Type in any non digit character e.g. whitespaces, letters or special characters
6. Verify none of the typed characters are entered in the search box
7. Type in any digit
8. Verify the typed digits are entered in the search box, upto 4 digits

**Comment by s77rt** [comment]
@trjExpensify, @hayata-suenaga did request a change on April 13th so technically it was not 100% ready for merge on April 13th.

**Comment by trjExpensify** [comment]
Yeah, because the changes were still applied within the timeframe and no further changes were required after that, the delay from the 13th onwards were on us I think.

**Comment by trjExpensify** [comment]
Okay offers have been sent.

**Comment by trjExpensify** [comment]
Settled up! 

@s77rt re: the regression test, there's this one [here](https://expensify.testrail.io/index.php?/cases/view/2458007&group_by=cases:section_id&group_order=asc&display_deleted_cases=0&group_id=229064) in TestRail. Instead of a totally separate one wouldn't we just add to that test script for ease? 

Maybe.. 

8. .... 
9. Type in any non digit character e.g. whitespaces, letters or special characters
10. **Verify** you see an error
11. ....

**Comment by s77rt** [comment]
@tjferriss I feel the test won't fit well with those since this one requires attention to the year selection page which none of existing tests mention.

Actually I think we can skip the regression test for this one just to be consist with the types of regression tests on that page.

**Comment by hayata-suenaga** [comment]
> The offending PR has been commented on: I don't think this applies here since it's not a regression. cc @hayata-suenaga

I agree

**Comment by trjExpensify** [comment]
>this is on hold for payment

Payment has been made, we're just looking at this regression test. 

>I feel the test won't fit well with those since this one requires attention to the year selection page which none of existing tests mention.

Ah, wait. The test script has changed since I linked it - likely because it has been recently developed. Okay, so there's this [standalone test for DOB](
https://expensify.testrail.io/index.php?/cases/view/2458041&group_by=cases:section_id&group_id=229064&group_order=asc&display_deleted_cases=0) now. 

So what we could do is insert these two lines before 5 & 6 perhaps? 

5. Type in any non digit character e.g. whitespaces, letters or special characters
6. **Verify** you see an error

**Comment by s77rt** [comment]
@trjExpensify Yes that looks good to me, but step 6 is to verify that the non of the typed characters got inserted (there is no error message)

**Comment by trjExpensify** [comment]
Sorry, I interpreted [this suggested script](https://github.com/Expensify/App/issues/17280#issuecomment-1526631864) as being that these characters are allowed for the input but receive an error on blur (a bit like forms). I just checked this and it's a numerical keyboard, so I think we can forgo this addition to the script - it's extraneous. 

Going to close this out!

**Comment by s77rt** [comment]
@trjExpensify You can still enter any char in Web / Desktop. But I think we can still skip this test.
